index.js

```diff
@@ -110,13 +110,18 @@ function attachFinishedListener (msg, callback) {
     if (finished) return
     if (eeMsg !== eeSocket) return
 
+    var attached = socket.__onFinished
+    if (!attached || !attached.queue) {
+      attached = socket.__onFinished = createListener(socket)
+      eeSocket = first([[socket, 'error', 'close']], attached)
+    }
     // finished on first socket event
-    eeSocket = first([[socket, 'error', 'close']], onFinish)
+    attached.queue.push(onFinish)
   }
 
   if (msg.socket) {
     // socket already assigned
-    onSocket(msg.socket)
+    defer(onSocket, msg.socket)
```

Comment:
Note: this change was needed to pass this test, since now the reused socket will already have an error/close listener. 
https://github.com/jshttp/on-finished/blob/0d88438c80e400db3af4a79d95408baa2f9f3f1b/test/test.js#L283-L295

The `.destroy()` is causing the callback to fire in the same tick. This might have highlighted an issue with the test. But the `defer` may also be adding handlers to a stale socket 🤔

Reply:
So that statement got me curious, and I pulled down your PR locally and reverted all the changes to `index.js` and the test failed, as expected. I then _only_ put back this line, the `defer` line, and then the tests passed. So it would seem like this one line is the only thing that makes the test pass; all the listening queue changes above don't seem to be contributing to the test passing. Are you able to replicate that?

Reply:
haha 🎉! that makes sense and I can confirm. If we give it the current tick to process onFinish for established sockets.. problem solved.
I've updated the PR!

Reply:
Ah, so should this use `process.nextTick` instead of `defer` then? The `defer` variable is an alias to `setImmediate` which is typically much more than a single tick in the event loop.

Reply:
So nextTick will put it in the same event loop iteration I think, which doesn't work because its still called before the close/end callbacks.

Reply:
So I put a debugger on it, and it seems with the `defer` there, the reason the warning went away is because the requests are always finished by the time it goes to add the socket listener since in the test it sends a response with no content immediately. I am thinking that a normal web server would not typically do that. If I add a delay to the `res.end` in the test, then indeed the test is failing with only the `defer` there, my bad.

```patch
diff --git a/test/test.js b/test/test.js
index cef5bf6..c53b7c6 100644
--- a/test/test.js
+++ b/test/test.js
@@ -559,7 +559,7 @@ describe('onFinished(req, listener)', function () {
       var stderr = captureStderr()
       var server = http.createServer(function (req, res) {
         onFinished(req, noop)
-        res.end()
+        setTimeout(() => res.end(), 1000)
       })
       server.on('close', function () {
         assert.strictEqual(stderr(), '')
```

Reply:
That also makes sense... Added back the original 😅

Reply:
Ok, so back to digging into what this `defer` seems to be doing. I see that removing it makes an existing test fail, so it seems adding it is actually to make a test that was already here pass again after the listener queue made it fail. Debugging why the listener queue makes that test fail, it looks like it's because in the logic added, the `eeSocket` is changed over on the first add to the queue, and that is set to callback the queue. But once the first request completes, it calls `eeSocket.cancel()` which cancels ever calling it back, even though the queue is still there (since it was never invoked). This means that once the first request in the pipeline completes and cancels the socket listener, all future requests then will never get socket events since they keep pushing then listener to a queue that was canceled by the first one completing.

Adding the `defer` just happens to make that one test pass as it creating a timing condition to end up adding the socket listener after the request completes, therefore never canceling it out.

Reply:
Hah, Yeah I saw the same while I was grabbing lunch. I'll see what I can come up with